frontend/src/components/vue_canvas/coordinators/coordinator_types/box/BoxInstanceCoordinator.ts

```diff
@@ -0,0 +1,28 @@
+import {Coordinator} from "../../Coordinator";
+import {BoxInstance} from "../../../instances/BoxInstance";
+
+export class BoxInstanceCoordinator extends Coordinator{
```

Comment:
What if this component's chief aim is to implement variations of  a `next(Interaction, Event)`function?

The main idea is to invert triggers based on mouse up/down. Instead to merge those as just a reason to "rerun" the loop. And the loop makes it's own choices based on Interaction + New Event.

So instead of processing a mouse down, it is about doing a 'query' that based on a given set of Interaction + New Event, what concrete functions (if any) should be called.

The test is to avoid having any type of JS event concepts as the primary path.

I'll make a concrete example to try and illustrate idea

`mouse_up()`
calls `ellipse_mouse_up()`

What if instead
1) The "global" router, is responsible from going from any type of interaction (just mouseup, mousedown to start), to say the appropriate type, here `ellipseCordinator`
(For example because Interaction.active_tool = 'ellipse' and ABC etc)

2) `ellipseCoordinator` will check a variety of conditions, to see if something should happen. For example, consider this existing function (slimmed to show point) 

```
ellipse_mouse_up():
      if (!this.ellipse_current_drawing_face && this.draw_mode) {
        this.$store.commit("init_draw");
      } else {
        this.$store.commit("finish_draw");
      }
```

We are already conditioning on if we are in the drawing state or not, to declare if we are starting or stopping drawing. 
So the fact that it's a mouse up, perhaps doesn't even matter? This function could be called equally well for mouse up or mouse down.

So for that function (ignoring use of vuex), if we just renamed it `ellipse.determine_drawing_state()` or something (just spitballing name), it could be called by either mouse up or mouse down.

I'm sure in practice there's some rough edges to that, but the point being that we should never rely on mouseup/mousedown as the primary assumption about the meaning of the interaction routing.

More generally
`def next()` ,
is based on set of `interaction.ABC` properties, to check if any coordinator actions are mapped to a set of those properties.
Which could include `interaction.mouse_state = something` or various properties.

The query could just be if statements, but the idea is the interaction itself is driving it, the mouse down is just one of the pieces of information it can use.

Let me know what you think :)

Reply:
Thanks for sharing this example Anthony! I like the direction, one question though is where do you think this `next()` method is defined? I find the name a bit confusing since it feels like its trying to get the next event on the interaction but really the objective is to somehow consume the event and execute any actions based on the properties right? 

Right now I added the `process_mouse_down`, `process_mouse_up` and `process_mouse_move` to the Coordinator class. So the BoxCoordinator for example will process the mouse down like this:

```

  public process_mouse_down(): CoordinatorProcessResult {
    console.log('process_mouse_down BOX', this.box_instance)
    let result: CoordinatorProcessResult = {
      instance_moved: false
    }
    if(!this.box_instance && !this.annotation_event.is_actively_drawing){
      let currently_drawing_box: BoxInstance = this.annotation_event.current_drawing_instance as BoxInstance
      currently_drawing_box.set_actively_drawing(true)
      result.is_actively_drawing = true
    }

    return result;
  }
```

This is roughly similar to what you are proposing right? Just want to double check if this is the case or if maybe I'm missing something.

Reply:
Hey Pablo
On this coordinator class would be one option.
Some other naming ideas for it could be
* take_action_from_interaction
* map_interaction_to_functions
* get_next_concrete_function

The current implementation is starting with the js events, the mouse_up,down,move etc.
The proposal is to delete/remove those as the baseclass functions
and replace them with the above function, or something sort of in that direction

the "next" name may be bad, the theory was to signify that there is a series of events in the interaction log, and that it's taking the "next" step in the interaction, based on the new event. 

Let's discuss a bit on the call :)